src/crypto/CHIPCryptoPALmbedTLS.cpp

```diff
@@ -194,8 +194,11 @@ Hash_SHA256_stream::~Hash_SHA256_stream(void) {}
 
 static inline mbedtls_sha256_context * to_inner_hash_sha256_context(HashSHA256OpaqueContext * context)
 {
-    static_assert(sizeof(context->mOpaque) >= sizeof(mbedtls_sha256_context), "Need more memory for SHA256 Context");
+    static_assert(sizeof(HashSHA256OpaqueContext) >= sizeof(mbedtls_sha256_context), "Need more memory for SHA256 Context");
     static_assert(std::is_trivially_copyable<mbedtls_sha256_context>(), "mbedtls_sha256_context values must copyable");
+    // Some architectures like ARM generate a fault on unaligned memory access, so verify the alignment.
```

Comment:
The fact that, for portability, this still has to be verified safe at runtime is really concerning.

Why don't we build `HashSHA256OpaqueContext` as a union of all the things it may need to be cast to?

Reply:
I use `static_assert` so it's verified at compilation time. Or maybe I misunderstood you? @pan-apple may know better why we don't use a union of mbedtls/openssl context, but my guess is that he wanted `CHIPCryptoPAL.h` to be independent from specific crypto implementations.

Reply:
> I use `static_assert` so it's verified at compilation time.

I see.  That is better.

However, `P256KeypairContext` appears to still have the bug.  I think this demonstrates why this reinterpret_cast pattern is unsafe: it's error prone. 

If performance is key and copies must be avoided at all costs, we should do this right and create structures with which we can cast safely.

Reply:
You're right about `P256KeypairContext`. Thanks! I'll fix that one, too. 

I'm not sure how to better detect similar issues. We probably could introduce a helper like below and try to adopt it in some places, but I'm not sure if it will bring much value:
```
template<class T>
class align_aware_cast
{
  template<class U>
  T operator(U other)
  {
    static_assert(sizeof(T) >= sizeof(U), "Wrong size");
    static_assert(alignof(T) >= alignof(U), "Wrong alignment");
    return reinterpret_cast<T>(other);
  }
};
```
and try to adopt it in such places?

Reply:
I'm not sure either how to detect these issues, except to carefully scrutinize all uses of reinterpret_cast and avoid it where possible.

As to what to do here, as long as all of the `to_` and `from_` methods in `src/crypto` have the proper protections, I think that's fine.  However, counting them up and including all of the cases with `P256KeypairContext`, I count 9.

I think your template actually would have value and make this easier.

Reply:
Anything coming out of or going into network buffers or packed buffers should be using those methods in `src/lib/core/CHIPEncoding.h` or the `*Unaligned` methods in `third_party/nlio/repo/include/nlio-base.hpp` or various wrappers thereto in the project.

Reply:
@msandstedt I added the template and fixed `P256KeypairContext` as well.